src/format_macro.rs

```diff
@@ -0,0 +1,258 @@
+                symbols.simple_symbol("append"),
+                Default::default(),
+            )));
+
+            let format_string = match &arg.value {
+                ast::Expr::Literal(ast::Literal::String(text)) => text,
+                _ => {
+                    return Err(macros::Error::message(format!(
+                        "format! expects a string argument"
+                    )))
+                }
+            };
+
+            let show_expr = |e: ast::SpannedExpr<'ast, Symbol>| {
+                let func = show_func.ast_clone(arena.borrow());
+
+                sp(ast::Expr::App {
+                    func,
+                    implicit_args: arena.alloc_extend(vec![]),
+                    args: arena.alloc_extend(vec![e]),
+                })
+            };
+
+            let app_exprs = |lhs, rhs| {
+                let func = append_func.ast_clone(arena.borrow());
+
+                sp(ast::Expr::App {
+                    func,
+                    implicit_args: arena.alloc_extend(vec![]),
+                    args: arena.alloc_extend(vec![lhs, rhs]),
+                })
+            };
+
+            let literal_expr = |val: String| sp(ast::Expr::Literal(ast::Literal::String(val)));
+
+            let type_cache = TypeCache::new();
+
+            let mut remaining = format_string.as_str();
+
+            let mut result_expr = None;
+
+            while let Some(find_result) = find_expr(remaining)? {
+                remaining = find_result.remaining;
+
+                let sub_expr = parse_expr(arena.borrow(), symbols, &type_cache, find_result.expr)
+                    .map_err(|err| {
+                    macros::Error::message(format!(
+                        "format! could not parse subexpression: {}",
+                        err
+                    ))
+                })?;
+
+                let part_expr = app_exprs(
+                    literal_expr(find_result.prefix.to_owned()),
+                    show_expr(sub_expr),
+                );
+
+                result_expr = match result_expr.take() {
+                    None => Some(part_expr),
+                    Some(prev_expr) => Some(app_exprs(prev_expr, part_expr)),
+                };
+            }
+
+            let result_expr = match result_expr.take() {
+                None => literal_expr(remaining.to_owned()),
+                Some(result_expr) => {
+                    if remaining.is_empty() {
+                        result_expr
+                    } else {
+                        app_exprs(result_expr, literal_expr(remaining.to_owned()))
+                    }
+                }
+            };
+
+            Ok(result_expr.into())
+        })
+    }
+}
+
+const OPEN_BRACE: &'static str = "{";
+const CLOSE_BRACE: &'static str = "}";
+
+struct FindExprResult<'a> {
+    prefix: &'a str,
+    expr: &'a str,
+    remaining: &'a str,
+}
+
+fn find_expr<'a>(format_str: &'a str) -> Result<Option<FindExprResult<'a>>, macros::Error> {
```

Comment:
Allowing arbitrary expressions inside the braces has some more wrinkles than just counting the braces. Consider that braces may also appear inside a string or character

```f#
format! "test {x + '}'} abc"
```
This would pick the first closing brace as the closer whereas it ought to be the second.

That could perhaps be fixed by parsing as an expression after finding the `{`, however then there is another problem with line comments.

```f#
format! "test {x // print x }"
```
The comment probably shouldn't contain the `}` here but if we naively parse everything after the `{` it would be.

It may be better to just start by allowing single variables in the format string (and possible positional arguments?)

Reply:
This is a good point. We could forbid comments in the embedded expressions: try to parse starting from the '{' then walk the resulting expression and raise an error if a comment is found.

Alternatively, allowing single variables could be a fair compromise because it would be forward compatible with full expressions, if we allow capturing variables in local scope

```gluon
let x = 1
in format! "x is {x}"
```

instead of explicitly passing to the macro

```gluon
format! "x is {x}" {x = 1}
```

Thoughts on this?

As for positional arguments, I am not convinced they are worth having but I can add them if you disagree.

Reply:
> As for positional arguments, I am not convinced they are worth having but I can add them if you disagree.

True, though that would require some lexer tweaks since it will just ignore comments atm.

> Alternatively, allowing single variables could be a fair compromise because it would be forward compatible with full expressions, if we allow capturing variables in local scope

Yeah, :+1: .

> As for positional arguments, I am not convinced they are worth having but I can add them if you disagree.

Yeah, perhaps. Lets wait with positional as they complicates things in combination with captures.